executor/merge_join.go

```diff
@@ -33,27 +32,41 @@ type MergeJoinExec struct {
 	baseExecutor
 
 	stmtCtx         *stmtctx.StatementContext
-	prepared        bool
-	outerKeys       []*expression.Column
-	innerKeys       []*expression.Column
-	outerIter       *readerIterator
-	innerIter       *readerIterator
-	outerFilter     []expression.Expression
-	resultGenerator joinResultGenerator
 	compareFuncs    []chunk.CompareFunc
-	outerIdx        int
-	outerRows       []chunk.Row
-	innerRows       []chunk.Row
-	outerIter4Row   chunk.Iterator
-	innerIter4Row   chunk.Iterator
+	resultGenerator joinResultGenerator
+
+	prepared bool
+	outerIdx int
+
+	innerTable *mergeJoinInnerTable
+	outerTable *mergeJoinOuterTable
+
+	innerRows     []chunk.Row
+	innerIter4Row chunk.Iterator
+}
+
+type mergeJoinOuterTable struct {
+	reader Executor
+	filter []expression.Expression
+	keys   []*expression.Column
+
+	chk      *chunk.Chunk
+	selected []bool
+
+	iter *chunk.Iterator4Chunk
+	row  chunk.Row
+}
+
+func (t *mergeJoinOuterTable) init(chk *chunk.Chunk, maxChunkSize int) {
+	t.chk = chk
+	t.iter = chunk.NewIterator4Chunk(chk)
+	t.selected = make([]bool, 0, maxChunkSize)
 }
 
-// readerIterator represents a row block with the same join keys
-type readerIterator struct {
-	stmtCtx  *stmtctx.StatementContext
-	sctx     sessionctx.Context
+// mergeJoinInnerTable represents inner table for merge join, which returns a
```

Comment:
mergeJoinInnerTable indicates inner table of merge join, which returns a
batch of rows with the same join key.